src/ts_catalog/continuous_agg.c

```diff
@@ -1486,35 +1498,114 @@ ts_continuous_agg_bucket_width(const ContinuousAgg *agg)
 	return agg->data.bucket_width;
 }
 
+/* Calls one of time_bukcket_ng() versions depending on the arguments. */
+static Datum
+generic_time_bucket_ng(const ContinuousAggsBucketFunction *bf, Datum timestamp)
+{
+	if (strlen(bf->timezone) > 0)
+	{
+		return DirectFunctionCall3(ts_time_bucket_ng_timezone,
+								   IntervalPGetDatum(bf->bucket_width),
+								   timestamp,
+								   CStringGetTextDatum(bf->timezone));
+	}
+	else
+	{
+		return DirectFunctionCall2(ts_time_bucket_ng_timestamp,
+								   IntervalPGetDatum(bf->bucket_width),
+								   timestamp);
+	}
+}
+
+/*
+ * Same as compute_inscribed_bucketed_refresh_window() but for variable-sized
+ * buckets.
+ *
+ * The algorithm is simple:
+ *
+ * end = time_bucket(bucket_size, end)
+ *
+ * if(start != time_bucket(bucket_size, start))
+ *     start = time_bucket(bucket_size, start) + bucket_size
+ *
+ */
+void
+ts_compute_inscribed_bucketed_refresh_window_variable(int64 *start, int64 *end,
+													  const ContinuousAggsBucketFunction *bf)
+{
+	Datum start_old, end_old, start_new, end_new;
+
+	start_old = ts_internal_to_time_value(*start, TIMESTAMPOID);
+	end_old = ts_internal_to_time_value(*end, TIMESTAMPOID);
+
+	start_new = generic_time_bucket_ng(bf, start_old);
+	end_new = generic_time_bucket_ng(bf, end_old);
+
+	if (DatumGetTimestamp(start_new) != DatumGetTimestamp(start_old))
+	{
+		// AALEKSEEV TODO: should we account for the timezone here? (e.g. gaps in the time)
+		start_new = DirectFunctionCall2(timestamp_pl_interval,
+										start_new,
+										IntervalPGetDatum(bf->bucket_width));
+	}
+
+	*start = ts_time_value_to_internal(start_new, TIMESTAMPOID);
+	*end = ts_time_value_to_internal(end_new, TIMESTAMPOID);
+}
```

Comment:
The logic of this function is pretty similar of the circumscribed function below. What about simplify it creating an internal function passing an parameter indicating to calculate `start_new` or `end_new`???

Reply:
That's usually not such a good idea. You save 5-10 lines of code for sure, but also create a **dependency** of two similar but unrelated procedures. This has a tendency to end up as a single procedure with 10 if-else's inside in the long run, not to mention the fact that the comments explaining what are the bool arguments for turn into a mess.